### PR TITLE
apps: React-related fixes

### PR DESCRIPTION
**Pull request by mareklibra**
Part of #9263 PR-split, see individual commits for more info.

The DBus-related issue was observed in tests when working on #9263.

### CODE REVIEW

File: pkg/apps/application-list.jsx [hunk 1]
```diff
@@ -140,18 +140,20 @@ class ApplicationList extends React.Component {
...
         return (
             <table className={table_classes}>
                 <caption>
                     <table>
-                        <tr>
-                            <td><h2>{_("Applications")}</h2></td>
-                            <td>{refresh_progress}</td>
-                            <td>{refresh_button}</td>
-                        </tr>
+                        <tbody>
```
Comment (martinpitt) [inline pkg/apps/application-list.jsx]: Ugh, is a tbody really always required in React? That sounds strange, it's entirely optional in HTML and only really useful if one needs to shove in additional things next to the tr?

File: pkg/apps/application.jsx [hunk 2]
```diff
@@ -80,11 +80,11 @@ class Application extends React.Component {
 
             return description.map(paragraph => {
                 if (paragraph.tag == 'ul') {
-                    return <ul>{paragraph.items.map(item => <li>{item}</li>)}</ul>;
+                    return <ul key={paragraph}>{paragraph.items.map(item => <li key={item}>{item}</li>)}</ul>;
```
Comment (martinpitt) [inline pkg/apps/application.jsx]: This would  repeat the entire paragraph text in the HTML, which seems rather useless and wasteful? What is the purpose of all these keys?

**Comment by mareklibra** [comment]
It's a good practice to explicitly use the `<tbody>`. Discussed i.e. here: https://github.com/facebook/react/issues/5652

For that reason, I'll need to fix a few CSS selectors within `check-*` scripts, since the DOM structure is changed. ANyway, it's already done within #9263 , so will get into these "split" PRs when relevant.

File: pkg/apps/application.jsx (hunk 2 above)
Comment (mareklibra) [reply pkg/apps/application.jsx]: Discussed in #9339 , the long paragraph text will not make it to the HTML

**Comment by mareklibra** [comment]
Files under `pkg/apps` are modified only in this PR, so I think the storage fixes are unrelated.
For storage, some of the fixes will be just similar to this one.

File: pkg/apps/application.jsx (hunk 2 above)
Comment (mvollmer) [reply pkg/apps/application.jsx]: But `paragraph` is not a string...

File: pkg/apps/utils.jsx [hunk 3]
```diff
@@ -91,7 +91,7 @@ export const show_error = ex => {
     if (ex.code == "cancelled")
         return;
 
-    if (ex.code == "not-found")
+    if (ex.code == "not-found" || ex.problem == "not-found")
```
Comment (mvollmer) [inline pkg/apps/utils.jsx]: Can you explain this more?  `ex.code` is from PackageKit, and `ex.problem` is from cockpit-bridge.  The first is "not-found" when the package was not known to PackageKit, the latter is "not-found" when PackageKit itself isn't there.

Let's move this out into its own PR.

File: pkg/apps/application-list.jsx [hunk 4]
```diff
@@ -140,18 +140,20 @@ class ApplicationList extends React.Component {
             tbody = <tr className="app-list-empty"><td>{empty_caption}</td></tr>;
         } else {
             table_classes += " table-hover";
-            tbody = comps.map((c) => <ApplicationRow comp={c} />);
+            tbody = comps.map((c) => <ApplicationRow comp={c} key={c} />);
```
Comment (mvollmer) [inline pkg/apps/application-list.jsx]: `c` is not a string.  `c.id` should work.

File: pkg/apps/application.jsx [hunk 5]
```diff
@@ -62,7 +62,7 @@ class Application extends React.Component {
         function render_homepage_link(urls) {
             return urls.map(url => {
                 if (url.type == 'homepage') {
-                    return (<div className="app-links">
+                    return (<div className="app-links" key={url}>
```
Comment (mvollmer) [inline pkg/apps/application.jsx]: `url.link` should work

File: pkg/apps/application.jsx [hunk 6]
```diff
@@ -80,11 +80,11 @@ class Application extends React.Component {
 
             return description.map(paragraph => {
                 if (paragraph.tag == 'ul') {
-                    return <ul>{paragraph.items.map(item => <li>{item}</li>)}</ul>;
+                    return <ul key={paragraph}>{paragraph.items.map(item => <li key={item}>{item}</li>)}</ul>;
                 } else if (paragraph.tag == 'ol') {
-                    return <ol>{paragraph.items.map(item => <li>{item}</li>)}</ol>;
+                    return <ol key={paragraph}>{paragraph.items.map(item => <li key={item}>{item}</li>)}</ol>;
```
Comment (mvollmer) [inline pkg/apps/application.jsx]: Key is not a string.

File: pkg/apps/application.jsx [hunk 7]
```diff
@@ -127,7 +127,7 @@ class Application extends React.Component {
                     {render_homepage_link(comp.urls)}
                     <div className="app-description">{render_description(comp.description)}</div>
                     <center>
-                        { comp.screenshots.map(s => <img className="app-screenshot" role="presentation" src={s.full} />) }
+                        { comp.screenshots.map(s => <img key={s} className="app-screenshot" role="presentation" src={s.full} />) }
```
Comment (mvollmer) [inline pkg/apps/application.jsx]: Key is not a string.

File: pkg/apps/application.jsx [hunk 8]
```diff
@@ -80,11 +80,11 @@ class Application extends React.Component {
...
             return description.map(paragraph => {
                 if (paragraph.tag == 'ul') {
-                    return <ul>{paragraph.items.map(item => <li>{item}</li>)}</ul>;
+                    return <ul key={paragraph}>{paragraph.items.map(item => <li key={item}>{item}</li>)}</ul>;
                 } else if (paragraph.tag == 'ol') {
-                    return <ol>{paragraph.items.map(item => <li>{item}</li>)}</ol>;
+                    return <ol key={paragraph}>{paragraph.items.map(item => <li key={item}>{item}</li>)}</ol>;
                 } else {
-                    return <p>{paragraph}</p>;
+                    return <p key={paragraph}>{paragraph}</p>;
```
Comment (mvollmer) [inline pkg/apps/application.jsx]: Key is not a string.

File: pkg/apps/utils.jsx (hunk 3 above)
Comment (mareklibra) [reply pkg/apps/utils.jsx]: To be honest, I did not dig into details of this. Test was failing here and while debugging I watched for content of the `ex`. I will need to start over to give proper answer to your question.

File: pkg/apps/application.jsx (hunk 2 above)
Comment (mareklibra) [reply pkg/apps/application.jsx]: replacing with string `paragraph-${index}`. Using indexes is not preferred, but not anti-pattern and working.

File: pkg/apps/application-list.jsx (hunk 4 above)
Comment (mareklibra) [reply pkg/apps/application-list.jsx]: ok

File: pkg/apps/application.jsx (hunk 5 above)
Comment (mareklibra) [reply pkg/apps/application.jsx]: ok

File: pkg/apps/application.jsx (hunk 6 above)
Comment (mareklibra) [reply pkg/apps/application.jsx]: fixed via indexes. Or anything better (natural key)?

**Comment by mareklibra** [comment]
Fixed, can you have a look, please? I'm using array indexes to build the `key` string value. I believe, this might be ok in this case. If concerned, I can replace by either natural key (need to be found) or generally unique (using counter).

File: pkg/apps/utils.jsx (hunk 3 above)
Comment (mvollmer) [reply pkg/apps/utils.jsx]: Let's leave this out here, then.

**Comment by mvollmer** [comment]
React related changes look good to me now, thanks!

The D-Bus fix should not be in this PR.  It might or might not be a good fix, but it's unclear what it is fixing, and not related to React.

**Comment by mvollmer** [comment]
I have removed the D-Bus fix from this PR.
